src/paho/mqtt/subscribe.py

```diff
@@ -35,17 +37,19 @@ def _on_connect_v5(client, userdata, flags, rc, properties):
     else:
         client.subscribe(userdata['topics'], userdata['qos'])
 
+
 def _on_connect(client, userdata, flags, rc):
     """Internal v5 callback"""
     _on_connect_v5(client, userdata, flags, rc, None)
 
 
 def _on_message_callback(client, userdata, message):
     """Internal callback"""
-    userdata['callback'](client, userdata['userdata'], message)
+    userdata['callback'](client, userdata['userdata'],
```

Comment:
This is a breaking change. Existing user of subscribe.callback will be required to change the signature of their callback function to accept an additional parameter "lock".

The idea I see to avoid any breaking change and still support this feature (while simple() still just call callback()) would be to add support for a special exception "_StopSubscriber" that the user callback could raise to disconnect. The callback "_on_message_simple" could then raise that exception, this would continue to make simple() a normal user of callback().
So here we could do:
```
try:
  userdata['callback'](client, userdata['userdata'], message)
except _StopSubscriber:
  lock.release()
```

I suggest to name it "_StopSubscriber" because I would like to kept the existence of this special exception undocumented/unsupported for now, because I think the true solution would be to call "client.disconnect()" instead of this exception. But doing the disconnect will not work because currently the received message is acknowledged after the on_message callback so disconnecting from inside the callback will cause the acknowledge to be dropped.

Reply:
Ah, I thought that since it is an internal function I could change the signature.

I could very simply determine whether userdata['lock'] is set, and if not call with the original signature instead. This would keep it compatible and would simply add the new functionality.

Using an exception here is not something I'd like to do. An exception always models something that is not part of the normal execution path, and in addition, is normally quite a bit slower. I think the approach with checking the value of 'lock' would be better. Or is there another motivation for using the exception?

Reply:
It's not only an internal function or I did I missunderstood the code ? It's the callback passed to subscribe.callback() which is user provided.

In subscribe.simple() I agree it's internal, but we added timeout & the lock also on subscribe.callback().

I just realize we recently added a `user_data_get` on client. It's a bit broken in that use-case, since it will return the "internal" userdata instead of the user-userdata (`userdata["userdata"]`). But that probably solve our issue. We can access to `userdata["lock"]` by using `client.user_data_get()["lock"]`. This avoid any change in callback signature.